### DISCUSSION TITLE
KeyCloak Refresh External IDP Token

### DISCUSSION THREAD

**Discussion by ahcamy2021** [keycloak/keycloak#15938] [Q&A]
Hi,

We noticed that when the KeyCloak token is refreshed using "refresh_token", the tokens from the external IDP are not. 

We took Take a look at the code https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/broker/oidc/OIDCIdentityProvider.java#L186-L187, when you perform an external token exchange keycloak will refresh the external access token if it has expired and the external refresh token has not.

Is there a reason why Keycloak does not refresh External IDP Token? Or it's a bug?

Could you propose any workaround to refresh the tokens from the external IDP ?

Thanks,

**Comment by AndreasBBS** [comment]
I opened [this issue](https://github.com/keycloak/keycloak/issues/19599) to see if this situation is addressed. 

At the moment my workaround is to Enable the Request Refresh Token on the IDP and if the Token Exchange returns an expired token I [retrieve the IDP refresh token](https://www.keycloak.org/docs/latest/server_development/index.html#retrieving-external-idp-tokens) and get a new access token with it.

Hope this helps!
